pandas/core/dtypes/cast.py

```diff
@@ -253,72 +267,388 @@ def changeit():
+        The value that the output dtype needs to be able to hold.
+
+    Returns
+    -------
+    dtype : DType
+        The updated dtype.
+    fill_value : scalar
+        The type of this value depends on the type of the passed fill_value
+
+        * If fill_value is a scalar, the method returns that scalar, but
+          modified to fit the updated dtype. For example, a datetime fill_value
+          will be returned as an integer (representing ns) for M8[ns], and
+          values considered missing (see pd.isna) will be returned as the
+          corresponding missing value marker for the updated dtype.
+        * If fill_value is an ndarray/Series/Index, this method will always
+          return the missing value marker for the updated dtype. This value
+          will be None for dtypes that cannot hold missing values (integers,
+          booleans, bytes).
+
+    See Also
+    --------
+    maybe_promote_with_scalar : underlying method for scalar case
+    maybe_promote_with_array : underlying method for array case
+    """
+    if is_scalar(fill_value) or isinstance(fill_value, tuple):
+        return maybe_promote_with_scalar(dtype, fill_value)
+    elif isinstance(fill_value, (np.ndarray, ABCSeries, ABCIndexClass)):
+        return maybe_promote_with_array(dtype, fill_value)
+    else:
+        fill_type = type(fill_value)
+        raise ValueError('fill_value must either be scalar, or a Series / '
+                         'Index / np.ndarray; received {}'.format(fill_type))
+
+
+def maybe_promote_with_scalar(dtype, fill_value=np.nan):
+    """
+    Determine minimal dtype to hold fill_value, when starting from dtype
+
+    Parameters
+    ----------
+    dtype : DType
+        The dtype to start from.
+    fill_value : scalar or np.ndarray / Series / Index
+        The value that the output dtype needs to be able to hold.
+
+    Returns
+    -------
+    dtype : DType
+        The updated dtype.
+    fill_value : scalar
+        The passed fill_value, potentially modified to fit the updated dtype.
+        For example, assuming a datetime dtype, a datetime.datetime fill_value
+        will be returned as an integer (representing ns) for M8[ns]. Similarly,
+        values considered missing (see pd.isna) will be returned as the
+        corresponding missing value marker for the updated dtype.
+
+    See Also
+    --------
+    maybe_promote_with_array : similar method for array case
+        This method contains the actual promotion logic for both cases.
+
+    Examples
+    --------
+    >>> maybe_promote(np.dtype('int'), fill_value=np.nan)
+    (dtype('float64'), nan)
+    >>> maybe_promote(np.dtype('float'), fill_value='abcd')
+    (dtype('O'), 'abcd')
+
+    For datetimes, timedeltas and datetimes with a timezone, the missing value
+    marker is pandas._libs.tslibs.iNaT (== np.iinfo('int64').min):
+
+    >>> maybe_promote(np.dtype('datetime64[ns]'), fill_value=np.nan)
+    (dtype('<M8[ns]'), -9223372036854775808)
```

Comment:
will this doctest fail on a big-endian system?

Reply:
Probably, but I don't consider this critical.